README

			   DJabberd::Delivery::OfflineStorage::InMemory,
			   DJabberd::Delivery::OfflineStorage::SQLite - 
           Old style offline storage of Messages


    Copyright (c) 2007 Piers Harding.
    All rights reserved.
    This program is free software; you can redistribute it and/or modify
    it under the terms of either:

	a) the GNU General Public License as published by the Free
	Software Foundation; either version 1, or (at your option) any
	later version, or

	b) the "Artistic License".


Whats it for?
============

DJabberd::Delivery::OfflineStorage::(InMemory|SQLite) - is a simple Offline Storage
mechanism for DJabberd.  This handles offline storage of messages in accordance to
XEP-0203 and XEP-0160.

If a user is offline, configured packets (should really only be Message type but
you could add IQ - really think about that though), then this plugin picks up
the packets at the end of the delivery chain and stores them. When the user next
annouces their presence, offline storage is checked, and any stored messages are
flooded through.


Installation
============

In brief, the following should work on most systems:
  perl Makefile.PL
	make
	make test 
	make install


Configuraton
============

So you want offline storage - well add this to djabberd.conf:

    <VHost mydomain.com>

        [...]
        <Plugin DJabberd::Delivery::OfflineStorage::InMemoryOnly>
             Types Message
         </Plugin>
    </VHost>

 For InMemoryOnly based storage, and:

    <VHost mydomain.com>

        [...]
        <Plugin DJabberd::Delivery::OfflineStorage::SQLite>
           Database offline.sqlite
           Types Message
        </Plugin>
    </VHost>
 For SQLite based storage.

Parameter Database specifies the sqlite database file to use, and Types is a list of Stanza
types that will be collected.  This should really only be Message - think really hard before
you add IQ to the list.

Also - it is extremely IMPORTANT as Edward Rudd pointed out - this Plugin MUST be the last
in the delivery chain, as it assumes that if S2S and Local haven't dealt with it, then the
JID in question is offline.


Cheers - Piers Harding - R/2 R/3 BASIS, IFACE, JAPH, whatever ...

email: piers@cpan.org




Changes
0.02 Thu Aug  2 18:38:01 BST 2007

Updated documentation to remind people that this must
be the last Plugin in the chain (thanks Edward Rudd)

0.03 Thu Nov 22 10:39:24 NZDT 2007
In memory messages not deleted - thanks  Michael Scherer

0.04 Tue Feb  5 19:36:09 NZDT 2008
Error with InMemoryStorage - moved to using full hash refs:
Error example for reference -
26238 INFO  DJabberd.Delivery.OfflineStorage.InMemoryOnly InMemoryOnly                                   
OfflineStorage load for: a@test.com                                                                      
  ... ERROR: [Can't call method "downbless" on an undefined value at  

0.05 Mon Dec  5 21:51:28 CET 2016
Added Presence stanza support
Added formal compliance with XEP-0203 (delay element)
Added formal compliance with XEP-0160 (disco#info and message filter)


lib/DJabberd/Delivery/OfflineStorage/SQLite.pm
package DJabberd::Delivery::OfflineStorage::SQLite;
use strict;
use base 'DJabberd::Delivery::OfflineStorage';
use warnings;
use DBI;

use vars qw($VERSION);
$VERSION = '0.05';

our $logger = DJabberd::Log->get_logger();

sub set_config_database {
    my ($self, $dbfile) = @_;
    $self->{dbfile} = $dbfile;
}

sub blocking { 1 };

sub finalize {
    my $self = shift;

    $logger->error_die("No 'Database' configured'") unless $self->{dbfile};

    my $dbh = DBI->connect_cached("dbi:SQLite:dbname=$self->{dbfile}","","",
                                  { RaiseError => 1, PrintError => 0, AutoCommit => 1 });
    $self->{dbh} = $dbh;
    $self->check_install_schema;
    $logger->info("Loaded SQLite OfflineStorage using file '$self->{dbfile}'");

    $self->SUPER::finalize;
}


sub load_offline_messages {
    my ($self, $user) = @_;

    # collect packets for user
    my @packets = ();
    eval {
      my $sth = $self->{dbh}->prepare("SELECT id, jid, timestamp, packet FROM offline WHERE jid=? ORDER BY id");
      $sth->execute($user);
      while (my $packet = $sth->fetchrow_hashref()){
        push(@packets, $packet);
      }
    };
    if ($@) {
      $logger->warn("SELECT against user '$user' on dbfile '$self->{dbfile}' failed with: $@");
    }
    return \@packets;
}


sub delete_offline_message {
    my ($self, $id) = @_;

    # delete packet by id
    eval {
      $self->{dbh}->do("DELETE FROM offline WHERE id=?", undef, $id);
    };
    if ($@) {
      $logger->warn("DELETE for '$id' on dbfile '$self->{dbfile}' failed with: $@");
    }
}


sub store_offline_message {
    my ($self, $user, $packet) = @_;

    eval {
      $self->{dbh}->do("INSERT INTO offline (jid, timestamp, packet) VALUES (?,datetime('now'),?)",
                       undef, $user, $packet);
    };
    if ($@) {
      $logger->warn("INSERT for user '$user' on dbfile '$self->{dbfile}' failed with: $@");
    }
}

sub check_install_schema {
    my $self = shift;
    my $dbh = $self->{dbh};

    eval {
        $dbh->do(qq{
            CREATE TABLE offline (
                                 id        INTEGER PRIMARY KEY AUTOINCREMENT,
                                 jid       VARCHAR(255) NOT NULL,
                                 timestamp VARCHAR(255) NOT NULL,
                                 packet    BLOB,
                                 UNIQUE (jid, id)
                                 )})
        };
    if ($@ && $@ !~ /table \w+ already exists/) {
        $logger->logdie("SQL error $@");
    }

}


1;


lib/DJabberd/Delivery/OfflineStorage/InMemoryOnly.pm
package DJabberd::Delivery::OfflineStorage::InMemoryOnly;
use strict;
use base 'DJabberd::Delivery::OfflineStorage';
use warnings;
use Data::Dumper;

use vars qw($VERSION);
$VERSION = '0.05';


our $logger = DJabberd::Log->get_logger();


sub load_offline_messages {
    my ($self, $user) = @_;
    $logger->info("InMemoryOnly OfflineStorage load for: $user");
    $self->{'offline'} ||= {};
    if (exists $self->{'offline'}{$user}) {
        my @messages = ();
        foreach my $message (sort keys %{$self->{'offline'}{$user}}) {
          push(@messages, $self->{'offline'}{$message});
        }
        return \@messages;
    } else {
        return [];
    }
}


sub delete_offline_message {
    my ($self, $id) = @_;
    $self->{'offline'} ||= {};
    $logger->info("InMemoryOnly OfflineStorage delete for: $id");
    # must remove it from $user too
    if (exists $self->{'offline'}->{$id}){
      my $user = $self->{'offline'}->{$id}->{'jid'};
      if (exists $self->{'offline'}->{$user}) {
        delete $self->{'offline'}->{$user}->{$id};
        delete $self->{'offline'}->{$user}
          unless keys %{$self->{'offline'}->{$user}};
      }
      delete $self->{'offline'}->{$id} 
    }
}


sub store_offline_message {
    my ($self, $user, $packet) = @_;
    $self->{offline} ||= {};
    $self->{offline_id} ||= 1;

    my $id = $self->{'offline_id'}++;
    $logger->info("InMemoryOnly OfflineStorage store for: $user/$id");
    $self->{'offline'}->{$user} ||= {};
    $self->{'offline'}->{$id} = {'id' => $id, 'packet' => $packet, 'jid' => $user};
    $self->{'offline'}->{$user}->{$id} = 1;
}

1;


lib/DJabberd/Delivery/OfflineStorage.pm
package DJabberd::Delivery::OfflineStorage;
use strict;
use warnings;
use base 'DJabberd::Delivery';

use DJabberd::Queue::ServerOut;
use DJabberd::Log;
our $logger = DJabberd::Log->get_logger;
use Storable qw(nfreeze thaw);
use POSIX qw(strftime);

use vars qw($VERSION);
$VERSION = '0.05';

=head1 NAME

DJabberd::Delivery::OfflineStorage - Basic OfflineStorage (XEP-0203) for DJabberd

=head1 DESCRIPTION

So you want offline storage - well add this to djabberd.conf:

    <VHost mydomain.com>

        [...]
        <Plugin DJabberd::Delivery::OfflineStorage::InMemoryOnly>
             Types Message
         </Plugin>
    </VHost>

 For InMemoryOnly based storage, and:

    <VHost mydomain.com>

        [...]
        <Plugin DJabberd::Delivery::OfflineStorage::SQLite>
           Database offline.sqlite
           Types Message
        </Plugin>
    </VHost>
 For SQLite based storage.

Parameter Database specifies the sqlite database file to use, and Types is a list of Stanza
types that will be collected.  This should really only be Message - think really hard before
you ad IQ to the list.

Also - it is extremely IMPORTANT as Edward Rudd pointed out - this Plugin MUST be the last
in the delivery chain, as it assumes that if S2S and Local haven't dealt with it, then the
JID in question is offline.


=head1 AUTHOR

Piers Harding, piers@cpan.org.


=cut



sub run_after { qw(DJabberd::Delivery::Local DJabberd::Delivery::S2S) }

sub package { __PACKAGE__ }

sub set_config_types {
    my ($self, $types) = @_;
    $self->{types} = { map { lc($_) => 1 }  grep(/^(IQ|Message)$/i, split(/\s+/, $types)) };
}


sub new {
    my $class = shift;
    my $self = $class->SUPER::new(@_);
    return $self;
}


sub register {
  my ($self, $vhost) = @_;
  $self->set_vhost($vhost);
  $vhost->register_hook("OnInitialPresence", sub { $self->on_initial_presence(@_) });
  $vhost->add_feature('msgoffline');
  $self->SUPER::register($vhost);
}


# OIntialPresence is used to determine that a user is now available
# and can receive stored offline messages
sub on_initial_presence {
    my ($self, $vhost, $cb, $conn) = @_;
    my $from = $conn->bound_jid
        or return;
    my $messages = $self->load_offline_messages($from->as_bare_string);
    # deliver messages
    foreach my $message (@$messages) {
      my $packet = Storable::thaw($message->{packet});
      my $class = $packet->{type};
      my $xml = DJabberd::XMLElement->new($packet->{ns}, $packet->{element}, $packet->{attrs}, []);
      my $stanza = $class->downbless($xml, $conn);
      $stanza->set_raw($packet->{stanza});
      $stanza->deliver($vhost);
      $self->delete_offline_message($message->{id});
    }
}

# Add XEP-203 delayed tag
sub delay {
    my ($time, $from, $text) = @_;
    my @text = ("$text") if($text);
    my %atts = (
	'{}xmlns' => "urn:xmpp:delay",
	'{urn:xmpp:delay}stamp' => strftime("%Y-%m-%dT%H:%M:%SZ",gmtime($time))
    );
    $atts{'{urn:xmpp:delay}from'} = $from if($from);
    return DJabberd::XMLElement->new('',"delay",\%atts,\@text);
}
sub add_delay {
    my ($stanza, $from, $text) = @_;
    $from = $stanza->connection->vhost->name if(!$from && $stanza->connection && $stanza->connection->vhost);
    my $delay = delay(time, $from, $text);
    $stanza->push_child($delay);
}

# hit the end of the delivery chain and we know that the user
# cannot accept the message -> store it offline for later
sub deliver {
    my ($self, $vhost, $cb, $stanza) = @_;
    die unless $vhost == $self->{vhost}; # sanity check

    my $to = $stanza->to_jid
        or return $cb->declined;

    # only configured packet types
    return $cb->declined
      unless exists $self->{types}->{$stanza->element_name};

    # additional filtering on message stanzas - only normal chats
    if($stanza->element_name eq 'message' and $stanza->attr('{}type')) {
        return $cb->declined unless($stanza->attr('{}type') eq 'normal' or
                ($stanza->attr('{}type') eq 'chat' and grep{$_->element_name eq 'body'}$stanza->children_elements));
    }

    add_delay($stanza,$vhost->name,"Offline delivery");
    my $packet = { 'type'    => ref($stanza),
                   'element' => $stanza->element_name,
                   'stanza'  => $stanza->innards_as_xml,
                   'ns'      => $stanza->namespace,
                   'attrs'   => {}                      };
    map { $packet->{attrs}->{$_} = $stanza->attrs->{$_} } keys %{$stanza->{attrs}};

    $self->store_offline_message($to->as_bare_string, Storable::nfreeze($packet));

    $DJabberd::Stats::counter{deliver_to_offline_storage}++;

    $cb->delivered;
    $logger->debug("Stashing stanza ".$stanza->as_xml);
}

=head1 COPYRIGHT & LICENSE

Original work Copyright 2006 Alexander Karelas, Martin Atkins, Brad Fitzpatrick and Aleksandar Milanov. All rights reserved.
Copyright 2007 Piers Harding.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut


1;
